lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,241 @@
+}) async {
+  var astVisitor = _LowerBoundConstraintVisitor(
+    rootPackage: rootPackageName,
+    warning: packageAnalysisContext.warning,
+    dependencySummaries: dependencySummaries,
+    targetDependencies: targetDependencies,
+    dependencyInstalledVersions: dependencyInstalledVersions,
+  );
+
+  final libPath = path.join(packageLocation, 'lib');
+  final libFolder = packageAnalysisContext.getFolder(libPath);
+
+  // retrieve the paths of all the dart library files in this package via the
+  // resourceProvider (.dart files in ./lib)
+  final dartLibFiles = getAllFiles(libFolder)
+      .where((file) => path.extension(file.path) == '.dart')
+      .map((file) => file.path)
+      .sorted();
+
+  for (final filePath in dartLibFiles) {
+    final result =
+        await packageAnalysisContext.analysisSession.getResolvedUnit(filePath);
+    if (result is ResolvedUnitResult) {
+      astVisitor.visitCompilationUnit(result.unit);
+    } else {
+      packageAnalysisContext.warning(
+          'Attempting to get a resolved unit resulted in an invalid result.');
+    }
+  }
+
+  return astVisitor.issues.values.whereNotNull().toList();
+}
+
+class _LowerBoundConstraintVisitor extends GeneralizingAstVisitor {
+  /// Maps from [Element.id] to either [LowerBoundConstraintIssue] if the lower
+  /// bound constraint on [LowerBoundConstraintIssue.dependencyPackageName] is
+  /// too low, making [Element] when the lowest allowed version
+  /// [LowerBoundConstraintIssue.lowestVersion] is used, or null otherwise.
+  final Map<int, LowerBoundConstraintIssue?> issues = {};
+
+  /// The name of the package being analyzed. Invocations corresponding to
+  /// definitions within this package will be ignored.
+  final String? rootPackage;
+
+  /// The summaries of each of the dependencies of the target package.
+  final Map<String, PackageShape> dependencySummaries;
+
+  /// The dependencies of the target package, as parsed from its pubspec.
+  final Map<String, HostedDependency> targetDependencies;
+
+  /// The versions of the dependencies of the target package, as installed.
+  final Map<String, Version> dependencyInstalledVersions;
+
+  /// Log a warning that something unexpected happened.
+  final void Function(String message) warning;
+
+  _LowerBoundConstraintVisitor({
+    required this.rootPackage,
+    required this.warning,
+    required this.dependencySummaries,
+    required this.targetDependencies,
+    required this.dependencyInstalledVersions,
+  });
+
+  // TODO: consider FunctionReference, PropertyAccess
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    // an invocation of a top-level function or a class method
+    super.visitMethodInvocation(node);
+
+    final element = node.methodName.staticElement;
+
+    if (element == null) {
+      warning(
+          'Element associated with identifier ${node.methodName.name} could not be resolved.');
+      return;
+    }
```

Comment:
```suggestion
    if (element == null) {
      // We cannot statically resolve what is invoked when a method is called on a variable with
      // dynamic type. In this case we just do nothing, since we can't know what is being called.
      return;
    }
```

Reply:
No need to print a warning, since this is an expect case that can naturally happen.